collector/mdadm_linux.go

```diff
@@ -18,199 +18,141 @@ package collector
 
 func init() {
 	Factories["mdadm"] = NewMdadmCollector
 }
 
-func evalStatusline(statusline string) (active, total, size int64, err error) {
-	matches := statuslineRE.FindStringSubmatch(statusline)
-
-	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
-	if len(matches) < 3+1 {
-		return 0, 0, 0, fmt.Errorf("too few matches found in statusline: %s", statusline)
-	} else if len(matches) > 3+1 {
-		return 0, 0, 0, fmt.Errorf("too many matches found in statusline: %s", statusline)
-	}
-
-	size, err = strconv.ParseInt(matches[1], 10, 64)
-	if err != nil {
-		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+func max(a, b int32) int32 {
+	if a > b {
+		return a
 	}
-
-	total, err = strconv.ParseInt(matches[2], 10, 64)
-	if err != nil {
-		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
-	}
-	active, err = strconv.ParseInt(matches[3], 10, 64)
-	if err != nil {
-		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
-	}
-
-	return active, total, size, nil
-}
-
-func evalRaid0line(statusline string) (size int64, err error) {
-	matches := raid0lineRE.FindStringSubmatch(statusline)
-
-	if len(matches) < 2 {
-		return 0, fmt.Errorf("invalid raid0 status line: %s", statusline)
-	}
-
-	size, err = strconv.ParseInt(matches[1], 10, 64)
-	if err != nil {
-		return 0, fmt.Errorf("%s in statusline: %s", err, statusline)
-	}
-
-	return size, nil
+	return b
 }
 
-func evalUnknownPersonalitylineRE(statusline string) (size int64, err error) {
-	matches := unknownPersonalityLineRE.FindStringSubmatch(statusline)
+// getArrayInfo gets RAID info via the GET_ARRAY_INFO IOCTL syscall.
+func getArrayInfo(md *mdStatus) (err error) {
+	// needs md.name as input
+	var (
+		array   mdu_array_info
+		devsize uint64
+	)
 
-	if len(matches) != 2+1 {
-		return 0, fmt.Errorf("invalid unknown personality status line: %s", statusline)
-	}
+	md.disksActive = math.NaN()
+	md.disksFailed = math.NaN()
+	md.disksMissing = math.NaN()
+	md.disksSpare = math.NaN()
+	md.bytesTotal = math.NaN()
+	md.bytesSynced = math.NaN()
+	md.active = math.NaN()
 
-	size, err = strconv.ParseInt(matches[1], 10, 64)
+	mdDev := "/dev/" + md.name
+	file, err := os.Open(mdDev)
 	if err != nil {
-		return 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+		return fmt.Errorf("error opening %s: %s", mdDev, err)
 	}
-
-	return size, nil
-}
-
-// evalBuildline gets the size that has already been synced out of the sync-line.
-func evalBuildline(buildline string) (int64, error) {
-	matches := buildlineRE.FindStringSubmatch(buildline)
-
-	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
-	if len(matches) < 1+1 {
-		return 0, fmt.Errorf("too few matches found in buildline: %s", buildline)
+	fd := file.Fd() // get the unix descriptor
+	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, fd, GET_ARRAY_INFO, uintptr(unsafe.Pointer(&array)))
```

Comment:
IMHO it feels a little weird to do manual syscall work in `node_exporter` itself.  It feels like there should be a separate `mdadm` package that the exporter consumes, so there is a clear separation of concerns.